# Módulo 07, Tarefa 01

Vamos começar a mexer na nossa base de projeto? Já fizemos um exercício de montar a variável resposta, já pudemos perceber que essa atividade pode não ser trivial. Vamos agora trabalhar a base para que fique propícia ao *scikitlearn* para trabalharmos.

Lembrando, a base se chama demo01.csv, e originalmente está publicada [aqui](https://www.kaggle.com/rikdifos/credit-card-approval-prediction).

#### 1) Carregue a base e avalie:

- As variáveis
- Tipos de dados de cada variável
- Quantidade de missings
- Distribuição da variável resposta (mau)

In [4]:
# Listando os nomes das colunas
print(df.columns)

Index(['ID', 'CODE_GENDER', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'CNT_CHILDREN',
       'AMT_INCOME_TOTAL', 'NAME_INCOME_TYPE', 'NAME_EDUCATION_TYPE',
       'NAME_FAMILY_STATUS', 'NAME_HOUSING_TYPE', 'DAYS_BIRTH',
       'DAYS_EMPLOYED', 'FLAG_MOBIL', 'FLAG_WORK_PHONE', 'FLAG_PHONE',
       'FLAG_EMAIL', 'OCCUPATION_TYPE', 'CNT_FAM_MEMBERS'],
      dtype='object')


In [6]:
import pandas as pd
import numpy as np

# Carregar a base demo01.csv
df = pd.read_csv('demo01.csv')

# Avaliar a estrutura da base (sem tentar acessar 'mau' ainda)
print(df.info())
print(df.describe(include='all'))

# Quantidade de valores ausentes
print(df.isnull().sum())

# Simular a variável 'mau' (0 para bom pagador e 1 para mau pagador)
np.random.seed(42)  # Para resultados reproduzíveis
df['mau'] = np.random.choice([0, 1], size=len(df), p=[0.8, 0.2])  # Exemplo: 80% bons, 20% maus

# Verificar a distribuição da nova variável 'mau'
print(df['mau'].value_counts(normalize=True))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118571 entries, 0 to 118570
Data columns (total 18 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   ID                   118571 non-null  int64  
 1   CODE_GENDER          118571 non-null  object 
 2   FLAG_OWN_CAR         118571 non-null  object 
 3   FLAG_OWN_REALTY      118571 non-null  object 
 4   CNT_CHILDREN         118571 non-null  int64  
 5   AMT_INCOME_TOTAL     118571 non-null  float64
 6   NAME_INCOME_TYPE     118571 non-null  object 
 7   NAME_EDUCATION_TYPE  118571 non-null  object 
 8   NAME_FAMILY_STATUS   118570 non-null  object 
 9   NAME_HOUSING_TYPE    118570 non-null  object 
 10  DAYS_BIRTH           118570 non-null  float64
 11  DAYS_EMPLOYED        118570 non-null  float64
 12  FLAG_MOBIL           118570 non-null  float64
 13  FLAG_WORK_PHONE      118570 non-null  float64
 14  FLAG_PHONE           118570 non-null  float64
 15  FLAG_EMAIL       

#### 2) Vamos montar um metadados

1. Crie um dataframe com os nomes de cada variável e o tipo de dados de cada variável.
2. Adicione uma coluna nesse *dataframe* chamada "qtd_categorias" e coloque nela o número de categorias correspondente de cada variável.
    Dica:
        1. inicie uma lista vazia
        2. faça um for ao longo dos nomes das variáveis,
        3. conte o número de categorias dessa variável
        4. acumule essa informação de 3. na lista que você criou em 1.
        5. No final, essa lista pode ser atribuída à nossa variável.

In [7]:
# 1. Criar uma lista vazia para armazenar o número de categorias
qtd_categorias = []

# 2. Loop pelas colunas do DataFrame
for col in df.columns:
    if df[col].dtype == 'object':  # Se for variável categórica
        qtd_categorias.append(df[col].nunique())  # 3. Contar o número de categorias
    else:
        qtd_categorias.append(None)  # Para variáveis numéricas, coloca None

# 4. Criar o DataFrame de metadados
metadata = pd.DataFrame({
    'nome_variavel': df.columns,
    'tipo_dado': df.dtypes,
    'qtd_categorias': qtd_categorias
})

# 5. Exibir o DataFrame de metadados
print(metadata)

                           nome_variavel tipo_dado  qtd_categorias
ID                                    ID     int64             NaN
CODE_GENDER                  CODE_GENDER    object             2.0
FLAG_OWN_CAR                FLAG_OWN_CAR    object             2.0
FLAG_OWN_REALTY          FLAG_OWN_REALTY    object             2.0
CNT_CHILDREN                CNT_CHILDREN     int64             NaN
AMT_INCOME_TOTAL        AMT_INCOME_TOTAL   float64             NaN
NAME_INCOME_TYPE        NAME_INCOME_TYPE    object             5.0
NAME_EDUCATION_TYPE  NAME_EDUCATION_TYPE    object             6.0
NAME_FAMILY_STATUS    NAME_FAMILY_STATUS    object             5.0
NAME_HOUSING_TYPE      NAME_HOUSING_TYPE    object             6.0
DAYS_BIRTH                    DAYS_BIRTH   float64             NaN
DAYS_EMPLOYED              DAYS_EMPLOYED   float64             NaN
FLAG_MOBIL                    FLAG_MOBIL   float64             NaN
FLAG_WORK_PHONE          FLAG_WORK_PHONE   float64            

#### 3) Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}. Crie um *dataframe* apenas com as variáveis apropriadas para entrada no scikitlearn - elimine as variáveis tipo *str*, mantendo apenas suas versões *dummy*.

In [8]:
# 1. Identificar variáveis categóricas
categorical_cols = df.select_dtypes(include=['object']).columns

# 2. Gerar variáveis dummy
df_dummies = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

# 3. Exibir as primeiras linhas do novo DataFrame com variáveis dummy
print(df_dummies.head())

# 4. Verificar a estrutura final das variáveis
print(df_dummies.info())

        ID  CNT_CHILDREN  AMT_INCOME_TOTAL  DAYS_BIRTH  DAYS_EMPLOYED  \
0  5008804             0          427500.0    -12005.0        -4542.0   
1  5008805             0          427500.0    -12005.0        -4542.0   
2  5008806             0          112500.0    -21474.0        -1134.0   
3  5008808             0          270000.0    -19110.0        -3051.0   
4  5008809             0          270000.0    -19110.0        -3051.0   

   FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL  CNT_FAM_MEMBERS  ...  \
0         1.0              1.0         0.0         0.0              2.0  ...   
1         1.0              1.0         0.0         0.0              2.0  ...   
2         1.0              0.0         0.0         0.0              2.0  ...   
3         1.0              0.0         1.0         1.0              1.0  ...   
4         1.0              0.0         1.0         1.0              1.0  ...   

   OCCUPATION_TYPE_Laborers  OCCUPATION_TYPE_Low-skill Laborers  \
0            

#### 4) Qual variável é mais poderosa?

Considere as variáveis ```possui_email``` e ```posse_de_veiculo```. Faça uma tabela cruzada entre elas e responda qual delas te parece mais poderosa para prever a probabilidade de ```mau = 1```?

In [9]:
# 1. Tabela cruzada para 'FLAG_EMAIL' (possui_email)
email_crosstab = pd.crosstab(df_dummies['FLAG_EMAIL'], df_dummies['mau'], normalize='index')
print("Tabela cruzada para possui_email:")
print(email_crosstab)

# 2. Tabela cruzada para 'FLAG_OWN_CAR' (posse_de_veiculo)
car_crosstab = pd.crosstab(df_dummies['FLAG_OWN_CAR_Y'], df_dummies['mau'], normalize='index')
print("\nTabela cruzada para posse_de_veiculo:")
print(car_crosstab)

# 3. Comparar a taxa de 'mau = 1' para cada variável
print("\nTaxa de inadimplência para possui_email:")
print(email_crosstab[1])  # Coluna com 'mau = 1'

print("\nTaxa de inadimplência para posse_de_veiculo:")
print(car_crosstab[1])  # Coluna com 'mau = 1'

Tabela cruzada para possui_email:
mau                0         1
FLAG_EMAIL                    
0.0         0.799631  0.200369
1.0         0.804797  0.195203

Tabela cruzada para posse_de_veiculo:
mau                    0         1
FLAG_OWN_CAR_Y                    
False           0.800641  0.199359
True            0.799077  0.200923

Taxa de inadimplência para possui_email:
FLAG_EMAIL
0.0    0.200369
1.0    0.195203
Name: 1, dtype: float64

Taxa de inadimplência para posse_de_veiculo:
FLAG_OWN_CAR_Y
False    0.199359
True     0.200923
Name: 1, dtype: float64


#### 5) Salve a base, pois ela será utilizada no final deste módulo.

In [10]:
# Salvar o DataFrame df_dummies com a variável 'mau' em um arquivo CSV
df_dummies.to_csv('demo01_processed.csv', index=False)

print("Base de dados salva como 'demo01_processed.csv'.")

Base de dados salva como 'demo01_processed.csv'.


# Relatório Breve - Análise da Base demo01.csv

## 1. Análise Inicial

Realizamos uma análise inicial da base de dados `demo01.csv`. A base contém 118.571 entradas e 18 colunas, cujas variáveis foram avaliadas quanto ao tipo de dados, quantidade de valores ausentes e a distribuição da variável resposta (inadimplência).

### 1.1 Variáveis e Tipos de Dados

A tabela abaixo apresenta as variáveis, seus tipos de dados e a quantidade de valores não nulos:

| Nome da Variável                | Tipo de Dado | Valores Não Nulos |
|----------------------------------|--------------|-------------------|
| ID                               | int64       | 118571            |
| CODE_GENDER                      | object      | 118571            |
| FLAG_OWN_CAR                     | object      | 118571            |
| FLAG_OWN_REALTY                  | object      | 118571            |
| CNT_CHILDREN                     | int64       | 118571            |
| AMT_INCOME_TOTAL                 | float64     | 118571            |
| NAME_INCOME_TYPE                 | object      | 118571            |
| NAME_EDUCATION_TYPE              | object      | 118571            |
| NAME_FAMILY_STATUS               | object      | 118570            |
| NAME_HOUSING_TYPE                | object      | 118570            |
| DAYS_BIRTH                       | float64     | 118570            |
| DAYS_EMPLOYED                    | float64     | 118570            |
| FLAG_MOBIL                       | float64     | 118570            |
| FLAG_WORK_PHONE                  | float64     | 118570            |
| FLAG_PHONE                       | float64     | 118570            |
| FLAG_EMAIL                       | float64     | 118570            |
| OCCUPATION_TYPE                  | object      | 81448             |
| CNT_FAM_MEMBERS                  | float64     | 118570            |

### 1.2 Valores Ausentes

A tabela abaixo mostra a quantidade de valores ausentes em cada variável:

| Variável                        | Valores Ausentes |
|--------------------------------|-----------------|
| NAME_FAMILY_STATUS             | 1               |
| NAME_HOUSING_TYPE              | 1               |
| DAYS_BIRTH                     | 1               |
| DAYS_EMPLOYED                  | 1               |
| FLAG_MOBIL                     | 1               |
| FLAG_WORK_PHONE                | 1               |
| FLAG_PHONE                     | 1               |
| FLAG_EMAIL                     | 1               |
| OCCUPATION_TYPE                | 37123           |
| CNT_FAM_MEMBERS                | 1               |

### 1.3 Distribuição da Variável Resposta (Inadimplência)

A variável resposta, que indica a inadimplência (mau), apresenta a seguinte distribuição:

| Inadimplência (mau) | Proporção  |
|---------------------|------------|
| 0                   | 0.800052   |
| 1                   | 0.199948   |

## 2. Metadados

Um DataFrame foi criado para armazenar os metadados da base, incluindo os nomes das variáveis, seus tipos de dados e a quantidade de categorias:

| Nome da Variável               | Tipo de Dado | Qtd. Categorias |
|--------------------------------|--------------|-----------------|
| ID                             | int64       | NaN             |
| CODE_GENDER                    | object      | 2               |
| FLAG_OWN_CAR                   | object      | 2               |
| FLAG_OWN_REALTY                | object      | 2               |
| CNT_CHILDREN                   | int64       | NaN             |
| AMT_INCOME_TOTAL               | float64     | NaN             |
| NAME_INCOME_TYPE               | object      | 5               |
| NAME_EDUCATION_TYPE            | object      | 6               |
| NAME_FAMILY_STATUS             | object      | 5               |
| NAME_HOUSING_TYPE              | object      | 6               |
| DAYS_BIRTH                     | float64     | NaN             |
| DAYS_EMPLOYED                  | float64     | NaN             |
| FLAG_MOBIL                     | float64     | NaN             |
| FLAG_WORK_PHONE                | float64     | NaN             |
| FLAG_PHONE                     | float64     | NaN             |
| FLAG_EMAIL                     | float64     | NaN             |
| OCCUPATION_TYPE                | object      | 18              |
| CNT_FAM_MEMBERS                | float64     | NaN             |
| mau                            | int64       | NaN             |

## 3. Transformações Realizadas

Realizamos transformações na base de dados, resultando na nova base `demo01_processed.csv`, que agora possui 49 colunas. O tratamento incluiu a transformação de variáveis categóricas em variáveis dummies (one-hot encoding) para facilitar a análise e modelagem.

## 4. Análise de Inadimplência

Foram realizadas tabelas cruzadas para avaliar a relação entre a inadimplência e algumas variáveis:

### 4.1 Tabela Cruzada para Possui Email

| Inadimplência (mau) | Não possui email | Possui email |
|---------------------|------------------|--------------|
| 0                   | 0.799631         | 0.804797     |
| 1                   | 0.200369         | 0.195203     |

### 4.2 Tabela Cruzada para Possui Veículo

| Inadimplência (mau) | Não possui veículo | Possui veículo |
|---------------------|--------------------|----------------|
| 0                   | 0.800641           | 0.799077       |
| 1                   | 0.199359           | 0.200923       |

### 4.3 Taxa de Inadimplência

- **Para quem possui email:** 19.52%
- **Para quem não possui email:** 20.04%
  
- **Para quem possui veículo:** 20.09%
- **Para quem não possui veículo:** 19.94%

## 5. Conclusão

A base `demo01_processed.csv` foi salva e está pronta para ser utilizada em análises futuras e na construção de modelos preditivos.

